### ISSUE TITLE
能否开放HTTPAdapter中max_retries参数，偶尔网络异常时，bucket的操作没有自动重试

### ISSUE THREAD

**Issue by wumanxing** [aliyun/aliyun-oss-python-sdk#86]
class Session(object):
    """属于同一个Session的请求共享一组连接池，如有可能也会重用HTTP连接。"""
    def __init__(self):
        self.session = requests.Session()

        psize = defaults.connection_pool_size
        self.session.mount('http://', requests.adapters.HTTPAdapter(pool_connections=psize, pool_maxsize=psize))
        self.session.mount('https://', requests.adapters.HTTPAdapter(pool_connections=psize, pool_maxsize=psize))

能否开放一下HTTPAdapter的配置，现在我除了改源码，没办法从外部修改重试次数，感谢各位大神

**Comment by yami** [comment]
会考虑增加配置项

**Comment by wumanxing** [comment]
在这之前,对于如何失败重试,能给个建议吗?必须要自己每个请求都包上重试代码?

**Comment by wumanxing** [comment]
@yami

**Comment by yami** [comment]
@wumanxing max_retries 也只能解DNS Lookup，连接超时等场景。对于上传、下载这种不一定是可以重试的，比如说上传的数据源可能是来自网络的。所以重试逻辑目前是交给了应用去做。应用程序可以抓住异常，根据HTTP Status和错误码进行一定次数的重试。比如说只重试500，RequestError等异常。

其他一些场景是SDK内部就可以重试的，比如说删除文件等。目前也没有特别的支持。

**Comment by wumanxing** [comment]
@yami  是的 如果是上传下载这种是应用控制比较合适的,但是像查询文件是否存在,或者查询信息,觉得可以在sdk里面内置一下,或者说开放一个参数.我们在生产环境遇到了这个问题,我本身可以通过代码实现重试,但是感觉这样做不太优雅.很冗余的感觉

**Comment by wumanxing** [comment]
@yami  而且这种重试只是为了解决偶然的网络问题. 这样如果网络情况特别差,重试还是失败,自然应用还是会处理的.所以如果sdk里面内置了重试,,对我们使用起来更方便了.毕竟像httpclient那种库都是可以配置自动重试的.

**Comment by yami** [comment]
@wumanxing 同意。这个工作我们需要安排一下。

**Comment by wumanxing** [comment]
@yami  非常感谢

**Comment by laixintao** [comment]
最好支持像requests那样可以传入session，或者对session mount一个用户自定义的Adapter
